larq_compute_engine/mlir/transforms/prepare_tf.cc

```diff
@@ -109,6 +112,38 @@ IntegerAttr GetNumChannels(Builder& b, Value output_val) {
   return b.getI32IntegerAttr(shape_vector[shape_vector.size() - 1]);
 }
 
+// Infers the number of groups from the shaped tensors of the input and the
+// filter. Will fail if any other type is passed. Will emit an error if the two
+// shapes are incompatible.
+IntegerAttr GetNumGroups(Builder& b, Value input_val, Value filter_val) {
+  auto input_type = input_val.getType().cast<ShapedType>();
+  auto input_shape_vector = input_type.getShape();
+  auto total_input_channels = input_shape_vector[input_shape_vector.size() - 1];
+  auto filter_type = filter_val.getType().cast<ShapedType>();
+  auto filter_shape_vector = filter_type.getShape();
+  auto filter_input_channels =
+      filter_shape_vector[filter_shape_vector.size() - 2];
+
+  auto num_groups = total_input_channels / filter_input_channels;
+  if (total_input_channels % filter_input_channels != 0) {
+    mlir::emitError(filter_val.getLoc())
+        << "Filter dimensions invalid: the number of filter input channels "
+        << filter_input_channels
+        << " does not divide the total number of input channels "
+        << total_input_channels << "\n";
+    num_groups = 1;
+  }
```

Comment:
Technically I don't think this is required, because the Conv2D op throws an error if this condition is violated, but it's probably better safe than sorry.